Adafruit_DAP_SAM.cpp

```diff
@@ -128,13 +130,101 @@ bool Adafruit_DAP_SAM::select(uint32_t *found_id) {
     if (device->dsu_did == DAP_DSU_did) {
       target_device = *device;
 
+      locked = dap_read_word(DAP_DSU_CTRL_STATUS) & 0x00010000;
+      if (locked) {
+        Serial.println("Device is locked, must be unlocked!");
+      }
+      else {
+        // Stop the core
+        finishReset();
+      }
       return true;
     }
   }
 
   return false;
 }
 
+void Adafruit_DAP_SAM::finishReset() {
+	// Stop the core
+	dap_write_word(DHCSR, 0xa05f0003);
+	dap_write_word(DEMCR, 0x00000001);
+	dap_write_word(AIRCR, 0x05fa0004);
+}
+
+void Adafruit_DAP_SAM::programFlash(const uint8_t * source, uint16_t bootloaderSize, bool doVerify, uint32_t flashOffset) {
+  size_t bufSize = pageSize();
+  uint8_t verifyBuf[bufSize];
+
+  Serial.println("\nProgramming...");
+  uint32_t startAddr = program_start(flashOffset);
+  uint32_t addr = startAddr;
+  while (addr < bootloaderSize) {
+    Serial.print(" 0x");
+    Serial.print(addr, HEX);
+    programBlock(addr, source + addr);
+    addr += bufSize;
+  }
+
+  if (doVerify) {
+    bool success = true;
+    Serial.println("\nVerifying...");
+
+    addr = startAddr;
+    while (addr < bootloaderSize) {
+      Serial.print("Block: 0x");
+      Serial.print(addr, HEX);
+      Serial.print(" ");
+      readBlock(addr, verifyBuf);
+      if (memcmp(verifyBuf, source + addr, bufSize) == 0) {
+        Serial.println("OK");
+      }
+      else {
+        Serial.println("Failed compare!");
+        success = false;
+      }
+      addr += bufSize;
+    }
+
+    if (!success) {
+      perror_exit("Error validating ");
+    }
+  }
+}
+
+void Adafruit_DAP_SAM::resetProtectionFuses(bool resetBootloaderProtection, bool resetRegionLocks) {
+  bool doFuseWrite = false;
+
+  fuseRead();
+
+  if (resetBootloaderProtection && _USER_ROW.bit.BOOTPROT != 7) {
+    Serial.print("Resetting BOOTPROT... ");
+    _USER_ROW.bit.BOOTPROT = 7;
+    doFuseWrite = true;
+  }
+  if (resetRegionLocks && _USER_ROW.bit.LOCK != 0xffffu) {
+    Serial.print(" Resetting NVM region LOCK... ");
+    _USER_ROW.bit.LOCK = 0xffffu;
+    doFuseWrite = true;
+  }
+
+  if (doFuseWrite) {
+    fuseWrite();
+  }
+}
+
+void Adafruit_DAP_SAM::resetWithExtension(void)
+{
+  Serial.print("Enter Reset with Extension mode... ");
+
+  // Enter reset extension mode
+  dap_reset_target_hw();
+  delay(10);
+  targetConnect();
+  Serial.println("Target prepare...");
+  dap_target_prepare();
```

Comment:
dap_target_prepare() shouldn't be part of this resetWithExtension(), we should try to have a primitive function here and leave the flow control to higher level.

https://github.com/ataradov/edbg/blob/master/edbg.c#L346

Reply:
The dap_target_prepare function needs to be called at the exactly the right places. It CANNOT be called before a reset with extension is called on a MCU that is protected with the Security Bit, and it MUST be called before using dap_write_word or dap_read_word.
So it basically needs to be called just after resetWithExtension has called targetConnect, before doing anything else. Every time. It may NOT be part of targetConnect, since it will fail when calling targetConnect on a protected MCU.

The edbg in effect calls its equivalent of dap_target_prepare, the dap_set_transfer_size method, in every dap_write_word/dap_read_word, just in case it wasn't done already. That is the other approach. It certainly isn't the job of the sketches or at a higher level.

So there really is just two possibilities: leave it as submitted in resetWithExtension and in the select methods, or remove it from them, and call it in the dap_write_word and dap_read_word, as edbg do it. 
Moving the call back into the sketches. would introduce the need for a lot of if proteced then do not call, if protected then reset extension and then call dap_target_prepare. Just a lot of repetetive, boilperplate code, that people doesn't understand the need for, when they flash an unprotected MCU.

Reply:
https://github.com/ataradov/edbg/blob/master/dap.c#L440
https://github.com/ataradov/edbg/blob/master/dap.c#L368

Reply:
@ksmith3036 Thank you again for detail explanation, I have overlooked this, indeed the way you are doing make lots more sense.